bin/testObservability/helper/helper.js

```diff
@@ -0,0 +1,987 @@
+        } else {
+          exports.debug(`EXCEPTION IN ${event_api_url !== exports.requestQueueHandler.eventUrl ? log_tag : 'Batch-Queue'} REQUEST TO TEST OBSERVABILITY : ${error.message || error}`, true, error);
+        }
+        exports.pending_test_uploads.count = Math.max(0,exports.pending_test_uploads.count - (event_api_url === 'api/v1/event' ? 1 : data.length));
+        return {
+          status: 'error',
+          message: error.message || (error.response ? `${error.response.status}:${error.response.statusText}` : error)
+        };
+      }
+    }
+  } else if (run >= 5) {
+    exports.debug(`EXCEPTION IN ${log_tag} REQUEST TO TEST OBSERVABILITY : Build Start is not completed and ${log_tag} retry runs exceeded`);
+    if(process.env.BS_TESTOPS_JWT != "null") exports.pending_test_uploads.count = Math.max(0,exports.pending_test_uploads.count-1);
+    return {
+      status: 'error',
+      message: 'Retry runs exceeded'
+    };
+  } else if(process.env.BS_TESTOPS_BUILD_COMPLETED !== "false") {
+    setTimeout(function(){ exports.uploadEventData(eventData, run+1) }, 1000);
+  }
+}
+
+exports.setTestObservabilityFlags = (bsConfig) => {
+  /* testObservability */
+  let isTestObservabilitySession = true;
+  if(bsConfig["testObservability"]) isTestObservabilitySession = ( bsConfig["testObservability"] == true );
+  if(process.env.BROWSERSTACK_TEST_OBSERVABILITY) isTestObservabilitySession = ( process.env.BROWSERSTACK_TEST_OBSERVABILITY == "true" );
+  if(process.argv.includes('--disable-test-observability')) isTestObservabilitySession = false;
+
+  /* browserstackAutomation */
+  let isBrowserstackInfra = true;
+  if(bsConfig["browserstackAutomation"]) isBrowserstackInfra = ( bsConfig["browserstackAutomation"] == true );
+  if(process.env.BROWSERSTACK_AUTOMATION) isBrowserstackInfra = ( process.env.BROWSERSTACK_AUTOMATION == "true" );
+  if(process.argv.includes('--disable-browserstack-automation')) isBrowserstackInfra = false;
+
+  process.env.BROWSERSTACK_TEST_OBSERVABILITY = isTestObservabilitySession;
+  process.env.BROWSERSTACK_AUTOMATION = isBrowserstackInfra;
+
+  return [isTestObservabilitySession, isBrowserstackInfra];
+}
+
+exports.isTestObservabilitySession = () => {
+  return ( process.env.BROWSERSTACK_TEST_OBSERVABILITY == "true" );
+}
+
+exports.isBrowserstackInfra = () => {
+  return ( process.env.BROWSERSTACK_AUTOMATION == "true" );
+}
+
+exports.shouldReRunObservabilityTests = () => {
+  return (process.env.BROWSERSTACK_RERUN_TESTS && process.env.BROWSERSTACK_RERUN_TESTS !== "null") ? true : false
+}
+
+exports.stopBuildUpstream = async (buildStartWaitRun = 0, testUploadWaitRun = 0, forceStop = false) => {
+  if(process.env.BS_TESTOPS_BUILD_COMPLETED !== "false" && !forceStop && exports.pending_test_uploads.count && testUploadWaitRun < 5) {
```

Comment:
This will be running in the cypress process and reporter will be running in the sdk process right? Will `pending_test_uploads.count` have a proper value here?

Reply:
Correct. Removing this check on stopBuild method.